installinator/src/reporter.rs

```diff
@@ -49,129 +78,418 @@ impl ProgressReporter {
+                        *peer,
+                        update_id,
+                        // Note: report_rx.clone() starts off as seen. However,
+                        // the initial state of report_rx is None, so if there's
+                        // an report in the receiver, it will always be treated
+                        // as unseen. This means that if a report is available,
+                        // the task will immediately dispatch it to the peer.
+                        report_rx.clone(),
+                        backend.clone(),
+                    );
+                    peer_tasks.insert(*peer, task);
+                }
             }
         })
     }
+}
 
-    async fn on_tick(&mut self) {
-        if let Some(task) = self.on_tick_task.take() {
-            // If the task we spawned on a previous tick is still running, just
-            // put it back; we'll wait for the next tick.
-            if !task.is_finished() {
-                self.on_tick_task = Some(task);
+async fn discover_task_loop(
+    log: &slog::Logger,
+    report_backend: ReportProgressBackend,
+    peers_tx: watch::Sender<Option<PeerAddresses>>,
+) {
+    let mut discover_interval = time::interval(Duration::from_secs(5));
+    // MissedTickBehavior::Skip ensures that discovery happens every
+    // interval, rather than bursting in case report_backend.discover_peers
+    // takes a long time.
+    discover_interval.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+    // Do not tick discover_interval immediately. We'd like one discovery to
+    // happen at the start of the loop.
+
+    loop {
+        discover_interval.tick().await;
+        let peers = match report_backend.discover_peers().await {
+            Ok(peers) => peers,
+            Err(DiscoverPeersError::Retry(error)) => {
+                slog::warn!(
+                    log,
+                    "failed to discover peers, will retry: {}",
+                    DisplayErrorChain::new(error.as_error()),
+                );
                 return;
             }
-
-            // Our last `on_tick` task is done; get its result, and possibly
-            // update `self.last_reported` (if we successfully reported
-            // progress to a peer).
-            if let Some(last_reported) = task.await.unwrap() {
-                self.last_reported = last_reported;
+            #[cfg(test)]
+            Err(DiscoverPeersError::Abort(_)) => {
+                // This is not possible, since test implementations don't
+                // currently generate Abort errors during reporter
+                // discovery.
+                unreachable!(
+                    "DiscoverPeersError::Abort is not generated for the \
+                    reporter by test implementations"
+                )
             }
+        };
+
+        slog::debug!(log, "discovered peers"; "peers" => ?peers);
+
+        if let Err(_) = peers_tx.send(Some(peers)) {
```

Comment:
Would `send_if_modified` be nicer here? Infallible and would avoid spurious wakeups of the reconciler task.

```rust
peers_tx.send_if_modified(|prev_peers| {
    if prev_peers.as_ref() == Some(&peers) {
        false
    } else {
        *prev_peers = Some(peers);
        true
    }
});

Reply:
thanks!